src/main.c

```diff
@@ -1,69 +1,123 @@
 #include <stdlib.h>
 #include <stdio.h>
 
-struct node {
-    struct node *pnext;
-    int item;
-};
+/* node data is stored here */
+#include "node.h"
 
-/*
 void setItem(struct node *node, int item) {
-    node -> item = item;
+    node->item = item;
 }
-*/
 
-/*
-void addNext(struct node *current, struct node next) {
+void addNext(struct node *current, struct node *next) {
     if(current -> pnext != NULL) {
-        current -> pnext = &next;
+        /**
+         * Simply taking a `struct node` and adding it with the dereference
+         * operator would not work here!
+         * This is because when you pass a struct node, it is passed by value.
+         * i.e. `next` would live in this functions stack.
+         *
+         * Therefore, when you assign `current->pnext = &next` it references
+         * this functions _own stack_ which will go away after the function
+         * returns.
+         *
+         * Therefore this function must take a `struct node *next` where the
+         * address is given, and set it to that.
+         */
+        current->pnext = next;
     } else {
-        //This doesn't behave how I expected it to
-        #warning attempting to add 'next' to a node that already has next
+        /**
+         * Preprocessor warnings occur whenever the preprocessor sees them. This
+         * occurs before compilation, and has nothing to do with actual source
+         * code. This is why preprocessor declarations should always be put at
+         * the beginning of a line.
+         * #warning creates a compilation warning—not a runttime one
+         */
+        fprintf(stderr, "error: "
+                "attempting to add 'next' to a node that already has next\n");
```

Comment:
Is there a way to check a condition and cause a compiler warning based on the result?

Reply:
No. the compiler only compiles your code. when your code is running the compiler is not involved.

There are some way to cause compile-time errors, but they should be used quite sparingly.